### PR TITLE
Adrienne / Added OIDC token and legacy token flow

### PR DESCRIPTION
**Pull request by adrienne-deriv**
### Motivation
- Added functions to integrate the OIDC access token retrieval flow and legacy token flow

### Actions

List the actions taken to achieve this change. What steps were followed? What was modified?

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [x] This change requires a documentation update

### CODE REVIEW

File: src/oidc/index.ts [hunk 1]
```diff
@@ -48,7 +53,76 @@ export const requestOidcAuthentication = async (
...
+        throw err
+    }
+}
+
+/**
+ * Fetches the tokens that will be passed to the `authorize` endpoint.
+ * 
+ * @param {string} accessToken The access token received after calling `requestOidcToken` successfully
+ */
+export const requestLegacyToken = async (accessToken: string) => {
```
Comment (thisyahlen-deriv) [inline src/oidc/index.ts:95]: Please add readme for these functions please

**Review by thisyahlen-deriv — APPROVED** [review]

**Review by hasan-deriv — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -120,3 +120,51 @@ const handleLoginClick = async () => {
...
+    const fetchToken = async () => {
+        const app_id = 'your-app-id'; // The ID of your app
+        const redirect_uri = 'http://your-app/callback'; // The URL to redirect to after successful login
+        const postLogoutRedirectUri = 'http://your-app/'; // The URL to redirect to after logging out
+
+        const { accessToken } = await requestOidcToken(app_id, redirect_uri, postLogoutRedirectUri)
+        // Once the access token is returned from `requestOidcToken` and is available, call `requestLegacyToken` to finally retrieve the tokens to pass into authorize
+        const tokens = await requestLegacyToken(accessToken);
+        // You can pass one of the tokens to authorize to login the user
+        callAuthorize(tokens.token1)
```
Comment (amam-deriv) [inline README.md:161]: [Q] will we receive multiple tokens for different loginid?

File: src/oidc/index.ts [hunk 3]
```diff
@@ -48,7 +53,76 @@ export const requestOidcAuthentication = async (
...
+            headers: {
+                'Authorization': `Bearer ${accessToken}`,
+            },
+        });
+        const data = await response.json()
+
+        return data
+    } catch (err) {
+        console.error("unable to request legacy tokens: ", err)
+        throw err
```
Comment (amam-deriv) [inline src/oidc/index.ts:110]: [M] throw new Error(), and is it possible to give a auth failure code to differentiate steps of failure?

File: src/oidc/index.ts [hunk 4]
```diff
@@ -48,7 +53,76 @@ export const requestOidcAuthentication = async (
...
+        const response = await fetch(`https://${server_url_from_local_storage}/oauth2/legacy/tokens`, { 
+            method : 'POST',
+            headers: {
+                'Authorization': `Bearer ${accessToken}`,
+            },
+        });
+        const data = await response.json()
+
+        return data
+    } catch (err) {
```
Comment (amam-deriv) [inline src/oidc/index.ts:108]: [M]
```suggestion
    } catch (errror) {
```

**Review by amam-deriv — APPROVED** [review]
LGTM!, though minor suggestions can be added next PR

File: src/oidc/index.ts [hunk 5]
```diff
@@ -9,13 +10,17 @@ type OidcConfiguration = {
...
 
+
 /**
  * Fetches the OIDC configuration for the given serverUrl.
  * @returns {Promise<object>} - A promise resolving to the OIDC configuration.
  * @throws {Error} - If there is a failure while fetching the OIDC configuration.
  */
 export const fetchOidcConfiguration = async (): Promise<OidcConfiguration> => {
-    const server_url_from_local_storage = localStorage.getItem('config.server_url') || '';
+    const config_from_local_storage = localStorage.getItem('config.oidc_endpoints')
```
[the remark this answers is not in the document]
Comment (adrienne-deriv) [reply src/oidc/index.ts:20]: No, the OIDC endpoints will be in one single source of truth which is the Well Known Configuration endpoint. All apps must query this endpoint to get the list of OIDC endpoints:

Screenshot 2024-10-28 at 10 40 21 AM

File: README.md (hunk 2 above)
Comment (adrienne-deriv) [reply README.md:161]: Yes, this is in the format of:
```json
{
   token1: ...
   acct1: ...
   token2: ...
   acct2: ...
}
```

Comment (adrienne-deriv) [reply README.md:161]: This is similar to what we have in our current authentication flow, which is the query parameters `?acct1=...,token1=...,acct2=...,token2=...`

File: src/oidc/index.ts (hunk 3 above)
Comment (adrienne-deriv) [reply src/oidc/index.ts:110]: I'm creating error types, it will be on a separate PR

**Review by thisyahlen-deriv — APPROVED** [review]

**Pull request merged**